### PR TITLE
feat: tx proposal

### PR DESCRIPTION
**Pull request by obiyankenobi**
I'm having some problems with `bitcore-lib` and `hathor-wallet-lib`. It'll be easier if I make some changes to the lib so I don't have to generate the addresses in this service and won't need to import `bitcore-lib` directly.

Since this code (address generation) was introduced on #4 , I'll go back to it and start making all the changes suggested. When that is done, I can come back to this PR to finish what's left. But in theory, most of the work is done and the tx is assembled and ready to be sent.

### CODE REVIEW

File: src/api/txProposalCreate.ts [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+  await addTxProposalOutputs(mysql, txProposalId, finalOutputs);
+
+  await closeDbConnection(mysql);
+
+  return {
+    statusCode: 200,
+    body: JSON.stringify({
+      success: true,
+      txProposalId,
+      inputs: inputUtxos.map((utxo) => ({ txId: utxo.txId, index: utxo.index })),
```
Comment (obiyankenobi) [inline src/api/txProposalCreate.ts]: Also need to return the tokens array, as it's needed to assemble the tx for mining.

File: .env.production [hunk 2]
```diff
@@ -1,3 +1,4 @@
 STAGE=production
-MAX_ADDRESS_GAP=20
+MAX_ADDRESS_GAP=10
```
Comment (msbrogli) [inline .env.production]: I feel the default value should be `20`. Any reason to change it?

File: .gitignore [hunk 3]
```diff
@@ -12,3 +12,5 @@ coverage
 
 # eslint
 .eslintcache
+
+.env
```
Comment (msbrogli) [inline .gitignore]: You removed the `.env` file. Shouldn't it be `.env.*`?

File: DATABASE.md [hunk 4]
```diff
@@ -35,6 +42,21 @@ CREATE TABLE `address_tx_history` (
   PRIMARY KEY (`address`,`tx_id`,`token_id`)
 );
 
+CREATE TABLE `version_data` (
```
Comment (msbrogli) [inline DATABASE.md]: Why don't we store it in memory? I feel we should get it from the fullnode when the service starts.

File: DATABASE.md [hunk 5]
```diff
@@ -48,19 +70,53 @@ CREATE TABLE `token` (
...
+  `id` varchar(36) NOT NULL,
+  `wallet_id` varchar(64) NOT NULL,
+  `status` enum('open','sent','send_error','cancelled') NOT NULL,
+  `created_at` int unsigned NOT NULL,
+  `updated_at` int unsigned DEFAULT NULL,
+  PRIMARY KEY (`id`)
+);
+
+
+CREATE TABLE `tx_proposal` (
```
Comment (msbrogli) [inline DATABASE.md]: Duplicate?

File: DATABASE.md [hunk 6]
```diff
@@ -48,19 +70,53 @@ CREATE TABLE `token` (
   `symbol` varchar(5) NOT NULL,
   PRIMARY KEY (`id`));
 
+CREATE TABLE `tx_proposal` (
+  `id` varchar(36) NOT NULL,
+  `wallet_id` varchar(64) NOT NULL,
+  `status` enum('open','sent','send_error','cancelled') NOT NULL,
+  `created_at` int unsigned NOT NULL,
+  `updated_at` int unsigned DEFAULT NULL,
+  PRIMARY KEY (`id`)
```
Comment (msbrogli) [inline DATABASE.md]: Maybe we should index `wallet_id` for fast queries.

File: DATABASE.md [hunk 7]
```diff
@@ -48,19 +70,53 @@ CREATE TABLE `token` (
...
+
+
+CREATE TABLE `tx_proposal_outputs` (
+  `tx_proposal_id` varchar(36) NOT NULL,
+  `index` tinyint unsigned NOT NULL,
+  `address` varchar(34) NOT NULL,
+  `token_id` varchar(64) NOT NULL,
+  `value` bigint DEFAULT NULL,
+  `timelock` int unsigned DEFAULT NULL,
+  PRIMARY KEY (`tx_proposal_id`,`index`)
```
Comment (msbrogli) [inline DATABASE.md]: In previous version, MySQL supported `O(1)` search using `tx_proposal_id` even though it is a multi-key hash index. Is it true for the latest version? Maybe we should include a `tx_proposal_id` hash index.

File: DATABASE.md [hunk 8]
```diff
@@ -48,19 +70,53 @@ CREATE TABLE `token` (
...
+);
+
+
+CREATE TABLE `tx_proposal_outputs` (
+  `tx_proposal_id` varchar(36) NOT NULL,
+  `index` tinyint unsigned NOT NULL,
+  `address` varchar(34) NOT NULL,
+  `token_id` varchar(64) NOT NULL,
+  `value` bigint DEFAULT NULL,
+  `timelock` int unsigned DEFAULT NULL,
```
Comment (msbrogli) [inline DATABASE.md]: Why do we need a timelock here? I feel we should never create a tx_proposal with locked outputs.

File: src/api/errors.ts [hunk 9]
```diff
@@ -8,9 +8,20 @@
 export enum ApiError {
   MISSING_PARAMETER = 'missing-parameter',
   INVALID_BODY = 'invalid-body',
+  INVALID_TX_WEIGHT = 'invalid-tx-weight',
+  INVALID_SELECTION_ALGO = 'invalid-selection-algo',
```
Comment (msbrogli) [inline src/api/errors.ts]: Why abbreviate?

File: src/api/wallet.ts [hunk 10]
```diff
@@ -1,20 +1,11 @@
-/**
```
Comment (msbrogli) [inline src/api/wallet.ts]: Removed the license header by mistake?

File: src/commons.ts [hunk 11]
```diff
@@ -0,0 +1,254 @@
+import { ServerlessMysql } from 'serverless-mysql';
```
Comment (msbrogli) [inline src/commons.ts]: Missing license header.

File: src/commons.ts [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+
+import {
```
Comment (msbrogli) [inline src/commons.ts]: Would you create a new PR for the refactor? It is way harder to review a mix of a big refactor with new features.

**Review by msbrogli — CHANGES_REQUESTED** [review]

**Comment by msbrogli** [comment]
@andreabadesso I haven't finished the review. I stopped when I reached a major refactor mixed with a new feature. I suggested you to split this PR in two parts: (i) a refactor, and (ii) the tx proposal feature.

File: DATABASE.md (hunk 4 above)
Comment (andreabadesso) [reply DATABASE.md]: Because the service is runnning on lambda, it should be stateless.

File: DATABASE.md (hunk 6 above)
Comment (andreabadesso) [reply DATABASE.md]: I agree. There is a separate issue to use an ORM and optimize the database though

File: DATABASE.md (hunk 8 above)
Comment (andreabadesso) [reply DATABASE.md]: This commit was killed in the rebase, but I believe you were asking about `timelock` in `tx_proposal_outputs`

`tx_proposal_outputs` are the outputs that are being created, not the utxos that are being spent

File: DATABASE.md (hunk 7 above)
Comment (andreabadesso) [reply DATABASE.md]: I agree, but there is already an issue to add indexes to tables

This will come with the ORM

File: DATABASE.md (hunk 8 above)
Comment (msbrogli) [reply DATABASE.md]: I got it. Thanks for the explanation.

File: src/api/txProposalCreate.ts (hunk 1 above)
Comment (andreabadesso) [reply src/api/txProposalCreate.ts]: Done in https://github.com/HathorNetwork/hathor-wallet-service/pull/36

**Pull request merged**